src/coreclr/vm/callstubgenerator.cpp

```diff
@@ -0,0 +1,1011 @@
+                                    break;
+                                case 16:
+                                    pInvokeFunction = CallJittedMethodRet4Float;
+                                    break;
+                                default:
+                                    _ASSERTE(!"Should not get here");
+                                    break;
+                            }
+                            break;
+                        case CORINFO_HFA_ELEM_DOUBLE:
+                            switch (thReturnValueType.GetSize())
+                            {
+                                case 8:
+                                    pInvokeFunction = CallJittedMethodRetDouble;
+                                    break;
+                                case 16:
+                                    pInvokeFunction = CallJittedMethodRet2Double;
+                                    break;
+                                case 24:
+                                    pInvokeFunction = CallJittedMethodRet3Double;
+                                    break;
+                                case 32:
+                                    pInvokeFunction = CallJittedMethodRet4Double;
+                                    break;
+                                default:
+                                    _ASSERTE(!"Should not get here");
+                                    break;
+                            }
+                                break;
+                        default:
+                            _ASSERTE(!"HFA types other than float and double are not supported yet");
+                            break;
+                    }
+                }
+                else
+                {
+                    switch (thReturnValueType.GetSize())
+                    {
+                        case 1:
+                        case 2:
+                        case 4:
+                        case 8:
+                            pInvokeFunction = CallJittedMethodRetI8;
+                            break;
+                        case 16:
+                            pInvokeFunction = CallJittedMethodRet2I8;
+                            break;
+                        default:
+                            _ASSERTE(!"The return types that are not HFA should be <= 16 bytes in size");
+                            break;
+                    }
+                }
+#else
+                _ASSERTE(!"Struct returns by value are not supported yet");
+#endif
+                break;
+            default:
+                _ASSERTE(!"Unexpected return type");
+                break;
+        }
+    }
+
+    m_routineIndex++; // Reserve one extra slot for the target method pointer
+
+    LoaderAllocator *pLoaderAllocator = pMD->GetLoaderAllocator();
+    S_SIZE_T finalStubSize(sizeof(CallStubHeader) + m_routineIndex * sizeof(PCODE));
+    void *pHeaderStorage = pLoaderAllocator->GetHighFrequencyHeap()->AllocMem(finalStubSize);
```

Comment:
We may want to check that some other thread have not created the stub in the meantime, and skip the rest if it did.

Reply:
The check happens at the caller where we store the stub to the MethodDescData.

Reply:
I meant we can reduce the amount of wasted loader heap allocations on race conditions if we recheck here.

Reply:
I think that if we used the AllocMemTracker as I've suggested in my other comment, we would not have this problem of leaking.